Web_server: prevent bootloop from css_include/js_include files being too large, also add option to set new index.html page for local: true

snetsplit opened the pull request:
# What does this implement/fix?

This fixes a bricking boot loop when js or css includes are too large. This pull also adds setting a custom local index html page with  a new option like:

```
 web_server:
  local: true
  port: 80
  local_page_include: "./index.html"
```

The behavior of the current release is the index html page is stored as a gzipped byte array statically, if config local: true. Presumably for speed and to save flash space. 

CSS, and JS includes, however, are stored as strings in RAM. The web_server loads at boot, and if either of these strings are too large, runs out of memory and crashes the esp system. This causes a boot loop.

This pull fixes that by instead reading CSS or JS includes, and then convert them to a gzipped byte array just like the site index page. Then cg.progmem_array is used to add them to c++ code as static PROGMEM constants where they can be accessed.  This saves flash memory and RAM memory and prevents the web_server component from crashing.

Further since this happens at compile time, it was simple to add an option for a custom local page include.

I'm unsure which file to update in the tests folder, but I will be very happy to if given guidance. Like wise this my first pull request and I hope everything is good, but please let me know if any changes need to happen. Thank you!

Edit: added tests to test3.yaml, and test7.yaml. Also in the test/components folder, added a web_server folder with test files. Hope this is what's being looked for.

## Types of changes

- [X] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes `[web_server bricks (boot loop) device if js_include is above 26KB(ish)](https://github.com/esphome/issues/issues/4303)`

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [x] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```
web_server:
  port: 80
  local: true
  custom_local_page: "./index.html"
  css_include: ./v2/www.css"
  js_include: "./v2/www.js"
  css_url: ""
  js_url: ""
```
## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

snetsplit committed:
Fix boot when js includes or css includes are too large, and implant a custom index page option

snetsplit committed:
Fix out of memory boot loop when js includes or css includes are too large, and implant a custom index page option

snetsplit committed:
Fix out of memory boot loop when js includes or css includes are too large, and implant a custom index page option

snetsplit committed:
Fix out of memory boot loop when js includes or css includes are too large, and implant a custom index page option

snetsplit committed:
Fix out of memory boot loop when js includes or css includes are too large, and implant a custom index page option

snetsplit committed:
Fix out of memory boot loop when js includes or css includes are too large, and implant a custom index page option

snetsplit committed:
Fix out of memory boot loop when js includes or css includes are too large, and implant a custom index page option

snetsplit committed:
Fix out of memory boot loop when js includes or css includes are too large, and implant a custom index page option

snetsplit committed:
Fix out of memory boot loop when js includes or css includes are too large, and implant a custom index page option

snetsplit committed:
Fix out of memory boot loop when js includes or css includes are too large, and implant a custom index page option

snetsplit committed:
Fix out of memory boot loop when js includes or css includes are too large, and implant a custom index page option

snetsplit committed:
Fix out of memory boot loop when js includes or css includes are too large, and implant a custom index page option

snetsplit committed:
Fix out of memory boot loop when js includes or css includes are too large, and implant a custom index page option

snetsplit committed:
Fix out of memory boot loop when js includes or css includes are too large, and implant a custom index page option

snetsplit committed:
Fix out of memory boot loop when js includes or css includes are too large, and implant a custom index page option

snetsplit committed:
Fix out of memory boot loop when js includes or css includes are too large, and implant a custom index page option

snetsplit committed:
Fix out of memory boot loop when js includes or css includes are too large, and implant a custom index page option

snetsplit committed:
Fix out of memory boot loop when js includes or css includes are too large, and implant a custom index page option

snetsplit committed:
Fix out of memory boot loop when js includes or css includes are too large, and implant a custom index page option

snetsplit committed:
Fix out of memory boot loop when js includes or css includes are too large, and implant a custom index page option

snetsplit committed:
Fix out of memory boot loop when js includes or css includes are too large, and implant a custom index page option

snetsplit committed:
Fix out of memory boot loop when js includes or css includes are too large, and implant a custom index page option

snetsplit committed:
Fix out of memory boot loop when js includes or css includes are too large, and implant a custom index page option

snetsplit committed:
Fix out of memory boot loop when js includes or css includes are too large, and implant a custom index page option

snetsplit committed:
Fix out of memory boot loop when js includes or css includes are too large, and implant a custom index page option

snetsplit committed:
Fix out of memory boot loop when js includes or css includes are too large, and implant a custom index page option

snetsplit committed:
Fix out of memory boot loop when js includes or css includes are too large, and implant a custom index page option

snetsplit committed:
Fix out of memory boot loop when js includes or css includes are too large, and implant a custom index page option

snetsplit committed:
Fix out of memory boot loop when js includes or css includes are too large, and implant a custom index page option

snetsplit committed:
Fix out of memory boot loop when js includes or css includes are too large, and implant a custom index page option

snetsplit committed:
Fix out of memory boot loop when js includes or css includes are too large, and implant a custom index page option

snetsplit committed:
Fix out of memory boot loop when js includes or css includes are too large, and implant a custom index page option

snetsplit committed:
Fix out of memory boot loop when js includes or css includes are too large, and implant a custom index page option

snetsplit committed:
Fix out of memory boot loop when js includes or css includes are too large, and implant a custom index page option

snetsplit committed:
Fix out of memory boot loop when js includes or css includes are too large, and implant a custom index page option

snetsplit committed:
Fix out of memory boot loop when js includes or css includes are too large, and implant a custom index page option

snetsplit committed:
Fix out of memory boot loop when js includes or css includes are too large, and implant a custom index page option

snetsplit committed:
Fix out of memory boot loop when js includes or css includes are too large, and implant a custom index page option

snetsplit committed:
Fix out of memory boot loop when js includes or css includes are too large, and implant a custom index page option

snetsplit committed:
Fix out of memory boot loop when js includes or css includes are too large, and implant a custom index page option

snetsplit committed:
Update index.html

snetsplit committed:
Merge branch 'dev' into dev

snetsplit committed:
Merge branch 'esphome:dev' into dev

snetsplit committed:
Merge branch 'esphome:dev' into dev

snetsplit committed:
Merge branch 'dev' into dev

snetsplit committed:
Merge branch 'dev' into dev

snetsplit committed:
make it so it can be imported as a component for testing

snetsplit committed:
make it so it can be imported as a component for testing without an updated const.py, and fix formatting complaints bylintc

snetsplit committed:
make it so it can be imported as a component for testing without an updated const.py, and fix formatting complaints bylintc

snetsplit committed:
Merge branch 'dev' into dev

snetsplit committed:
add climate controls to web_server

snetsplit committed:
Revert "add climate controls to web_server"

This reverts commit 58c741a78d3967c244ab3fbc13d634bc4dee2886.

snetsplit committed:
Merge branch 'esphome:dev' into dev

jesserockz commented on esphome/components/web_server/__init__.py:
```diff
@@ -17,7 +18,27 @@
 )
-from esphome.core import CORE, coroutine_with_priority
+
+# this import statement should make testing the PR much easier since othwerwise you get errors from const.py since it doesn't import with a custom component
+try:
+    from esphome.const import (
+        CONF_CSS_INCLUDE_ARRAY,
+        CONF_JS_INCLUDE_ARRAY,
+        CONF_LOCAL_PAGE_INCLUDE,
+        CONF_LOCAL_PAGE_INCLUDE_ARRAY,
+    )
+
+    globals()["CONF_CSS_INCLUDE_ARRAY"] = CONF_CSS_INCLUDE_ARRAY
+    globals()["CONF_JS_INCLUDE_ARRAY"] = CONF_JS_INCLUDE_ARRAY
+    globals()["CONF_LOCAL_PAGE_INCLUDE"] = CONF_LOCAL_PAGE_INCLUDE
+    globals()["CONF_LOCAL_PAGE_INCLUDE_ARRAY"] = CONF_LOCAL_PAGE_INCLUDE_ARRAY
+except ImportError:
+    globals()["CONF_JS_INCLUDE_ARRAY"] = "js_include_array"
+    globals()["CONF_CSS_INCLUDE_ARRAY"] = "css_include_array"
+    globals()["CONF_LOCAL_PAGE_INCLUDE"] = "local_page_include"
+    globals()["CONF_LOCAL_PAGE_INCLUDE_ARRAY"] = "local_page_include_array"
```
There is really no reason to add these fields to the common const.py file as they wont be used by other components.

snetsplit replied:
Thanks for the guidance! I've removed them const.py. Was trying to follow the existing patterns and wasn't sure if they needed to be in there for some reason.

jesserockz replied:
If you think a CONST will be reused, then for sure add to const.py, otherwise its fine in the component code. It can always be moved later.

snetsplit committed:
removing unneed constants from const.py

snetsplit committed:
removing unneed constants from const.py

snetsplit committed:
Merge branch 'dev' into dev

snetsplit committed:
removing unneed constants from const.py, and fixing lint-python formatting issue

snetsplit committed:
Merge branch 'dev' into dev

snetsplit committed:
Bug fix

CORS header is already added in web_server_base.h. Adding it in web_server.cpp just duplicates it, and breaks CORS: https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS/Errors/CORSMultipleAllowOriginNotAllowed

snetsplit committed:
Merge branch 'dev' into dev

snetsplit committed:
Merge branch 'dev' into dev

snetsplit committed:
Merge branch 'dev' into dev

snetsplit committed:
Merge branch 'dev' into dev

snetsplit committed:
Merge branch 'esphome:dev' into dev

snetsplit committed:
Merge branch 'esphome:dev' into dev
